# Tarea 5 - Aprendizaje Automático con Python

## Introducción

Para esta tarea realizaremos analisis de clasificación sobre la variable de Position (4). que cuenta con la sigientes clasificaciones: attack, goalkeeper, middle, defense. 
El objetivo general es ajustar un modelo de variables significativas que nos ayuden a clasificar en un porcentaje de certeza alto el valor de la variable de Position (4).

**Nota**: para la realizacion de la tarea 5 utilizaremos las siguientes librerias **numpy**, **pandas**, **sklearn**.

***

## Cargamos las librerias a utilizar

In [60]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import preprocessing
from sklearn import metrics
from sklearn.model_selection import train_test_split
%matplotlib inline


## Se carga el data set

In [61]:
path = '../dataframe/fifa_procesado.csv'
df = pd.read_csv(path)
len(df)

18207

## Vemos graficamente las tablas

In [62]:
df

Name  Age  \
0                L. Messi   31   
1       Cristiano Ronaldo   33   
2               Neymar Jr   26   
3                  De Gea   27   
4            K. De Bruyne   27   
...                   ...  ...   
18202        J. Lundstram   19   
18203  N. Christoffersson   19   
18204           B. Worman   16   
18205      D. Walker-Rice   17   
18206           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png       94         94   
1      https://cdn.sofifa.org/flags/38.png       94         94   
2      https://cdn.sofifa.org/flags/54.png       92         93   
3      https://cdn.sofifa.org/flags/45.png       91         93   
4       https://cdn.sofifa.org/flags/7.png       91         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

             Value  ...  LCB Grow  RCB Value RCB Grow  RB Value  RB Grow  \
0      110500000.0  ...       2.0       47.0      2.0      59.0      2.0   
1       77000000.0  ...       3.0       53.0      3.0      61.0      3.0   
2      118500000.0  ...       3.0       47.0      3.0      60.0      3.0   
3       72000000.0  ...       NaN        NaN      NaN       NaN      NaN   
4      102000000.0  ...       3.0       66.0      3.0      73.0      3.0   
...            ...  ...       ...        ...      ...       ...      ...   
18202      60000.0  ...       2.0       45.0      2.0      45.0      2.0   
18203      60000.0  ...       2.0       32.0      2.0      29.0      2.0   
18204      60000.0  ...       2.0       28.0      2.0      33.0      2.0   
18205      60000.0  ...       2.0       31.0      2.0      35.0      2.0   
18206      60000.0  ...       2.0       47.0      2.0      46.0      2.0   

       Work Rate Attack Work Rate Defense Position (4)   Position (13)  \
0                Medium            Medium       attack    right at

## Limpiamos el dataset de datos nulos y clomnuas que no seran consideradas en esete analisis

Limpiamos las columnas Work Rate Defense y Position (4) de filas con valores nulos

In [63]:
df = df[df['Work Rate Defense'].notna()]
df = df[df['Position (4)'].notna()]

Un paso importante es definir las variable **X** e **y** y que atributos o características la van a componer. Para ello se procede a visualizar las columnas del conjunto de datos.

In [97]:
print(df.columns)

Index(['Name', 'Age', 'Photo', 'Nationality', 'Flag', 'Overall', 'Potential',
       'Club', 'Club Logo', 'Value',
       ...
       'LCB Grow', 'RCB Value', 'RCB Grow', 'RB Value', 'RB Grow',
       'Work Rate Attack', 'Work Rate Defense', 'Position (4)',
       'Position (13)', 'Contract Length'],
      dtype='object', length=134)


Eliminamos columnas que no son relevantes para el analisis de clasificación 

In [91]:
#variables independientes
#['LS','ST','RS','LW','LF','CF','RF','RW','LAM','CAM','RAM','LM','LCM','CM','RCM','RM','LWB','LDM','CDM','RDM','RWB','LB','LCB','CB','RCB','RB']
datos = df.drop(['Position (13)','Name','Photo','Nationality','Flag','Club','Club Logo','Preferred Foot','Work Rate','Body Type','Real Face','Work Rate Attack', 'Position','LS','ST','RS','LW','LF','CF','RF','RW','LAM','CAM','RAM','LM','LCM','CM','RCM','RM','LWB','LDM','CDM','RDM','RWB','LB','LCB','CB','RCB','RB', 'Loaned From'],axis=1)
# Reemplazamos todos los valores nan por ceros 
datos = datos.fillna(0)
datos.head()


Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Jersey Number  ...  \
0                       5.0        4.0          4.0           10.0  ...   
1                       5.0        4.0          5.0            7.0  ...   
2                       5.0        5.0          5.0           10.0  ...   
3                       4.0        3.0          1.0            1.0  ...   
4                       4.0        5.0          4.0            7.0  ...   

   LB Grow  LCB Value  LCB Grow  RCB Value  RCB Grow  RB Value  RB Grow  \
0      2.0       47.0       2.0       47.0       2.0      59.0      2.0   
1      3.0       53.0       3.0       53.0       3.0      61.0      3.0   
2      3.0       47.0       3.0       47.0       3.0      60.0      3.0   
3      0.0        0.0       0.0        0.0       0.0       0.0      0.0   
4      3.0       66.0       3.0       66.0       3.0      73.0      3.0   

   Work Rate Defense  Position (4)  Contract Length  
0             Medium        attack       63158400.0  
1                Low        attack       94694400.0  
2             Medium        attack       94694400.0  
3             Medium    goalkeeper       31536000.0  
4               High        middle      126230400.0  

[5 rows x 94 columns]

Cambiamos el formato de la columna Work Rate Defense, utilizando el metodo get_dummies de pandas para volverla numerica

In [65]:
datos['Work Rate Defense'].unique()

array(['Medium', 'Low', 'High'], dtype=object)

In [92]:
wrd = pd.get_dummies(datos["Work Rate Defense"])

# combinar los dataframes "df" y "variable_dummy_1" 
datos = pd.concat([datos, wrd], axis=1)

# Elimina la columna original "fuel-type" del dataframe "df"
datos.drop("Work Rate Defense", axis = 1, inplace=True)
datos.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Jersey Number  ...  \
0                       5.0        4.0          4.0           10.0  ...   
1                       5.0        4.0          5.0            7.0  ...   
2                       5.0        5.0          5.0           10.0  ...   
3                       4.0        3.0          1.0            1.0  ...   
4                       4.0        5.0          4.0            7.0  ...   

   LCB Grow  RCB Value  RCB Grow  RB Value  RB Grow  Position (4)  \
0       2.0       47.0       2.0      59.0      2.0        attack   
1       3.0       53.0       3.0      61.0      3.0        attack   
2       3.0       47.0       3.0      60.0      3.0        attack   
3       0.0        0.0       0.0       0.0      0.0    goalkeeper   
4       3.0       66.0       3.0      73.0      3.0        middle   

   Contract Length  High  Low  Medium  
0       63158400.0     0    0       1  
1       94694400.0     0    1       0  
2       94694400.0     0    0       1  
3       31536000.0     0    0       1  
4      126230400.0     1    0       0  

[5 rows x 96 columns]

teniendo listo el datatset para el analisis creamos un nuevo csv del dataset de fifa

In [67]:
datos.to_csv('../dataframe/fifa_datos_limpios.csv')

Cargamos el dataset limpio para el analisis

In [128]:
path = '../dataframe/fifa_datos_limpios.csv'
df_entrenamiento_0 = pd.read_csv(path)
df_entrenamiento = df_entrenamiento_0.drop(['Unnamed: 0'], axis=1)
df_entrenamiento.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Jersey Number  ...  \
0                       5.0        4.0          4.0           10.0  ...   
1                       5.0        4.0          5.0            7.0  ...   
2                       5.0        5.0          5.0           10.0  ...   
3                       4.0        3.0          1.0            1.0  ...   
4                       4.0        5.0          4.0            7.0  ...   

   LCB Grow  RCB Value  RCB Grow  RB Value  RB Grow  Position (4)  \
0       2.0       47.0       2.0      59.0      2.0        attack   
1       3.0       53.0       3.0      61.0      3.0        attack   
2       3.0       47.0       3.0      60.0      3.0        attack   
3       0.0        0.0       0.0       0.0      0.0    goalkeeper   
4       3.0       66.0       3.0      73.0      3.0        middle   

   Contract Length  High  Low  Medium  
0       63158400.0     0    0       1  
1       94694400.0     0    1       0  
2       94694400.0     0    0       1  
3       31536000.0     0    0       1  
4      126230400.0     1    0       0  

[5 rows x 96 columns]

Obserbamos informacion sobre las columnas que utilizaremos

In [129]:
print(df_entrenamiento.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18147 entries, 0 to 18146
Data columns (total 96 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       18147 non-null  int64  
 1   Overall                   18147 non-null  int64  
 2   Potential                 18147 non-null  int64  
 3   Value                     18147 non-null  float64
 4   Wage                      18147 non-null  float64
 5   Special                   18147 non-null  int64  
 6   International Reputation  18147 non-null  float64
 7   Weak Foot                 18147 non-null  float64
 8   Skill Moves               18147 non-null  float64
 9   Jersey Number             18147 non-null  float64
 10  Joined                    18147 non-null  int64  
 11  Contract Valid Until      18147 non-null  int64  
 12  Height                    18147 non-null  float64
 13  Weight                    18147 non-null  float64
 14  Crossi

Tambien vemos cuántos hay en cada clase en el conjunto de datos.

In [130]:
df_entrenamiento['Position (4)'].value_counts()

middle        6838
defense       5866
attack        3418
goalkeeper    2025
Name: Position (4), dtype: int64

Hay 6838 jugadores en posición middle, 5866 jugadores en posición de defense, 3418 jugadores en posición de attack y 2025 jugadores en posición goalkeeper.

Se crea la variable df_y que contiene las posiciones de los jugadores y es la variable dependiente para el analisis

In [131]:
df_y = df_entrenamiento['Position (4)']
df_y

0            attack
1            attack
2            attack
3        goalkeeper
4            middle
            ...    
18142        middle
18143        attack
18144        attack
18145        attack
18146        middle
Name: Position (4), Length: 18147, dtype: object

Se define los valores de la variables independientes para el analisis 

In [132]:
df_x = df_entrenamiento.drop(['Position (4)'],axis=1)

In [133]:
X = df_x.values
X[0:5]

array([[3.1000000e+01, 9.4000000e+01, 9.4000000e+01, 1.1050000e+08,
        5.6500000e+05, 2.2020000e+03, 5.0000000e+00, 4.0000000e+00,
        4.0000000e+00, 1.0000000e+01, 1.5463008e+09, 1.6094592e+09,
        1.7676000e+02, 1.5900000e+02, 8.4000000e+01, 9.5000000e+01,
        7.0000000e+01, 9.0000000e+01, 8.6000000e+01, 9.7000000e+01,
        9.3000000e+01, 9.4000000e+01, 8.7000000e+01, 9.6000000e+01,
        9.1000000e+01, 8.6000000e+01, 9.1000000e+01, 9.5000000e+01,
        9.5000000e+01, 8.5000000e+01, 6.8000000e+01, 7.2000000e+01,
        5.9000000e+01, 9.4000000e+01, 4.8000000e+01, 2.2000000e+01,
        9.4000000e+01, 9.4000000e+01, 7.5000000e+01, 9.6000000e+01,
        3.3000000e+01, 2.8000000e+01, 2.6000000e+01, 6.0000000e+00,
        1.1000000e+01, 1.5000000e+01, 1.4000000e+01, 8.0000000e+00,
        2.2650000e+08, 8.8000000e+01, 2.0000000e+00, 8.8000000e+01,
        2.0000000e+00, 8.8000000e+01, 2.0000000e+00, 9.2000000e+01,
        2.0000000e+00, 9.3000000e+01, 2.0000000e

In [134]:
y = df_y
y[0:5]

0        attack
1        attack
2        attack
3    goalkeeper
4        middle
Name: Position (4), dtype: object

Creamos la data de entrenamiento y de pruebas para ver el rendimiento del modelo

In [135]:
X3_entrena, X3_prueba, y3_entrena, y3_prueba = train_test_split(X, y, test_size=0.3, random_state=3)
print ('Conjunto de Entrenamiento set:', X3_entrena.shape,  y3_entrena.shape)
print ('Conjunto de Prueba:', X3_prueba.shape,  y3_prueba.shape)

Conjunto de Entrenamiento set: (12702, 95) (12702,)
Conjunto de Prueba: (5445, 95) (5445,)


## Preparacion del modelo
El modelo a utlizar es el de Arboles de Decisión

In [136]:
from sklearn.tree import DecisionTreeClassifier

In [137]:
arbol_modelo = DecisionTreeClassifier(criterion="entropy", max_depth = 6)
arbol_modelo

DecisionTreeClassifier(criterion='entropy', max_depth=6)

Entrenamos el medelo con la data de entrenamiento

In [138]:
arbol_modelo.fit(X3_entrena,y3_entrena)

DecisionTreeClassifier(criterion='entropy', max_depth=6)

## Pronóstico

Probamos el modelo con el set de prueba y lo almacenamos para hacer la evaluación 

In [139]:
arbol_pronostico = arbol_modelo.predict(X3_prueba)

A simple vista se puede ver que para algunas clasificaciones acierta y en otras no.

In [140]:
print (arbol_pronostico [0:6])
print (y3_prueba [0:6])

['attack' 'attack' 'middle' 'attack' 'defense' 'middle']
5512      attack
3689      attack
3591      middle
2257      attack
473      defense
13232    defense
Name: Position (4), dtype: object


## Evalaución del modelo
Utilizando las métricas de **sklearn** y verifiquemos la precisión del modelo.

In [141]:
print("Precición del modelo basado en árbol de decisiones: ", metrics.accuracy_score(y3_prueba, arbol_pronostico))

Precición del modelo basado en árbol de decisiones:  0.8358126721763085


Como se puede ver el modelo tiene una precición sobre el 80% lo que nos da un indicador bastante realista de la clasificacion 